extlib/millhone/src/cmd/analyze_jars.rs

```diff
@@ -0,0 +1,188 @@
+struct JarAnalysis {
+    /// Jars and fingerprints associated with each layer in a jar file.
+    discovered_jars: HashMap<LayerPath, Vec<DiscoveredJar>>,
+}
+
+#[tracing::instrument]
+pub fn main(opts: Subcommand) -> Result<()> {
+    let tar_filename = opts.image_tar_file();
+    let jar_analysis = JarAnalysis {
+        discovered_jars: jars_in_container(opts.image_tar_file())
+            .with_context(|| format!("Analyze Jar Archive {:?}", tar_filename))?,
+    };
+
+    let mut stdout = BufWriter::new(std::io::stdout());
+    serde_json::to_writer(&mut stdout, &jar_analysis).context("Serialize Results")
+}
+
+/// Extracts the container (saved via `docker save`) and finds JAR files inside any layer.
+/// For each one found, fingerprints it and reports all those fingerprints along with their
+/// layer and path.
+#[tracing::instrument]
+fn jars_in_container(image_path: &PathBuf) -> Result<HashMap<LayerPath, Vec<DiscoveredJar>>> {
+    // Visit each layer and fingerprint the JARs within.
+    info!("inspecting container");
+    // May have to make two copies of this Archive, idk if it can be iterated twice.
+    let layers = list_container_layers(image_path)?;
+    let mut discoveries = HashMap::new();
+
+    let mut image = unpack(image_path)?;
+    for entry in image.entries().context("iterate entries")? {
+        let entry = entry.context("read entry")?;
+        let path = entry.path().context("read path")?;
+        if !layers.contains(path.as_ref()) {
+            debug!(?path, "skipped: not a layer file");
+            continue;
+        }
+
+        let layer = path.to_path_buf();
+        // Layers should have a form like blob
+        let layer_discoveries =
+            jars_in_layer(entry).with_context(|| format!("read layer '{layer:?}'"))?;
+        discoveries.insert(LayerPath(layer), layer_discoveries);
+    }
+
+    Ok(discoveries)
+}
+
+/// Open and unpack a file and put it into a tar.
+/// This is done repeatedly because `entries()` can only be read once from an Archive.
+#[tracing::instrument]
+fn unpack(path: &PathBuf) -> Result<Archive<File>> {
+    let file = File::open(path).context("open tar file")?;
+    Ok(tar::Archive::new(file))
+}
+
+#[tracing::instrument(skip(entry))]
+fn jars_in_layer(entry: Entry<'_, impl Read>) -> Result<Vec<DiscoveredJar>> {
+    let mut discoveries = Vec::new();
+
+    let mut entry_archive = tar::Archive::new(entry);
+    for entry in entry_archive.entries().context("list entries in layer")? {
+        let entry = entry.context("read entry")?;
+        let path = entry.path().context("read path")?;
+        if !path.to_string_lossy().ends_with(".jar") {
+            debug!(?path, "skipped: not a jar file");
+            continue;
+        }
+
+        let path = path.to_path_buf();
+
+        info_span!("jar", ?path).in_scope(|| -> Result<()> {
+            debug!("fingerprinting");
+            let entry = buffer(entry).context("Read jar file")?;
```

Comment:
[nit] due to the common ways errors are rendered, i recommend keeping the messages in `context` all lower case
```suggestion
            let entry = buffer(entry).context("read jar file")?;
```